pkg/controller/common/reconciler/reconciler.go

```diff
@@ -112,9 +112,24 @@ func ReconcileResource(params Params) error {
 			return err
 		}
 		return nil
+	}
+
+	// Check if already exists
+	err = params.Client.Get(types.NamespacedName{Name: name, Namespace: namespace}, params.Reconciled)
+	if err != nil && apierrors.IsNotFound(err) {
+		return create()
 	} else if err != nil {
 		log.Error(err, fmt.Sprintf("Generic GET for %s %s/%s failed with error", kind, namespace, name))
-		return errors.Wrapf(err, "failed to get %s %s/%s", kind, namespace, name)
+		return fmt.Errorf("failed to get %s %s/%s: %w", kind, namespace, name, err)
+	}
+
+	if params.NeedsDelete != nil && params.NeedsDelete() {
+		log.Info("Deleting resource", "kind", kind, "namespace", namespace, "name", name)
+		err := params.Client.Delete(params.Expected)
```

Comment:
We should apply delete preconditions here to ensure we don't re-delete the object we just recreated at the previous reconciliation.

```go

opt := client.Preconditions{
		UID:             reconciled.UID,
		ResourceVersion: reconciled.ResourceVersion,
}
err := k8sClient.Delete(reconciled, opt)
```

Reply:
Excellent point! I did not think of that.